### PR TITLE
Machinery for moving templates and generating hybrid recordings with drift.

### PR DESCRIPTION
**Pull request by samuelgarcia**
Depend on #2305  and #2317

Machinery for moving templates and generating hybrid recordings with drift.

Mainly include:
  * interpolate_templates(), move_dense_templates() : function for spatial interpolation
  * DriftingTemplates : extension of Templates class with displacement
  * InjectDriftingTemplatesRecording : similar than InjectTemplatesRecording but with drift

### CODE REVIEW

File: src/spikeinterface/core/template_tools.py [hunk 1]
```diff
@@ -248,3 +249,165 @@ def get_template_extremum_amplitude(
         unit_amplitudes[unit_id] = extremum_amplitudes[unit_id][best_channel]
 
     return unit_amplitudes
+
+
+def move_templates(templates, displacements, source_probe, dest_probe=None, interpolation_method="cubic"):
+    """
+    Move all templates given some displacements using spatial interpolation (cubic or linear).
+    Optionally can be remapped to another probe with a diffrents geometry.
```
Comment (zm711) [inline src/spikeinterface/core/template_tools.py:257]:
```suggestion
    Optionally can be remapped to another probe with a different geometry.
```

File: src/spikeinterface/core/template_tools.py [hunk 2]
```diff
@@ -248,3 +249,165 @@ def get_template_extremum_amplitude(
...
+
+    Before interpolation this function make is checking that templates can be interpolated on new
+    positions using the sparsity information : the entire sparse template must be covered at all new positions.
+    If the Templates object is dense then no check is possible and a wrning is emited.
+
```
Comment (zm711) [inline src/spikeinterface/core/template_tools.py:262]:
```suggestion
    Before interpolation this function checks that the templates can be interpolated to new
    positions using the sparsity information: the entire sparse templates must be covered at all new positions.
    If the Templates object is dense then no check is possible and a warning is emitted.

```

covered isn't quite the right word, but I'm still trying to think of a way to write this.

File: src/spikeinterface/core/template_tools.py [hunk 3]
```diff
@@ -248,3 +249,165 @@ def get_template_extremum_amplitude(
...
+        shape: (num_displacement, 2)
+    source_probe: Probe
+        The Probe object on which templates are defined
+    dest_probe: Probe or None
+        The destination Probe. Can be different geometry than the original.
+        If None then the same probe  is used.
+    interpolation_method: str, default "cubic"
+        The interpolation method.
+    interpolation_method: str, default "cubic"
+        The interpolation method.
```
Comment (zm711) [inline src/spikeinterface/core/template_tools.py:279]:
```suggestion
    dest_probe: Probe | None, default: None
        The destination Probe. Can be different geometry than the original.
        If None then the same probe  is used.
    interpolation_method: "cubic" | "linear",  default: "cubic"
        The interpolation method.
```

File: src/spikeinterface/core/template_tools.py [hunk 4]
```diff
@@ -248,3 +249,165 @@ def get_template_extremum_amplitude(
...
+
+def move_dense_templates(templates, displacements, source_probe, dest_probe=None, interpolation_method="cubic"):
+    """
+    Move all templates given some displacements using spatial interpolation (cubic or linear).
+    Optionally can be remapped to another probe with a diffrents geometry.
+
+    This function operate on dense template.
+
+    Note : in this function no check are done if template can be interpolable after displacements.
+    To make some check use teh higher level function move_templates().
```
Comment (zm711) [inline src/spikeinterface/core/template_tools.py:323]:
```suggestion
    Optionally can be remapped to another probe with a different geometry.

    This function operates on dense template.

    Note : in this function no checks are done to see if templates can be interpolatable after displacements.
    To check if the given displacements are interpolatable use the higher level function move_templates().
```

File: src/spikeinterface/core/template_tools.py [hunk 5]
```diff
@@ -248,3 +249,165 @@ def get_template_extremum_amplitude(
...
+        shape: (num_displacement, 2)
+    source_probe: Probe
+        The Probe object on which templates are defined
+    dest_probe: Probe or None
+        The destination Probe. Can be different geometry than the original.
+        If None then the same probe  is used.
+    interpolation_method: str, default "cubic"
+        The interpolation method.
+    interpolation_method: str, default "cubic"
+        The interpolation method.
```
Comment (zm711) [inline src/spikeinterface/core/template_tools.py:341]:
```suggestion
    dest_probe: Probe | None, default: None
        The destination Probe. Can be different geometry than the original.
        If None then the same probe  is used.
    interpolation_method: "cubic" | "linear", default: "cubic"
        The interpolation method.
```

File: src/spikeinterface/core/template_tools.py [hunk 6]
```diff
@@ -248,3 +249,165 @@ def get_template_extremum_amplitude(
...
+
+
+
+
+def interpolate_templates(templates, source_locations, dest_locations, interpolation_method="cubic"):
+    """
+    Interpolate template to a new position.
+    Uselfull to create motion or to remap template form probeA to probeB.
+
+    Note that several moves can be done by broadcating when dest_locations have one more dim.
```
Comment (zm711) [inline src/spikeinterface/core/template_tools.py:368]:
```suggestion
    Interpolate templates to new positions.
    Useful to create motion or to remap templates form probeA to probeB.

    Note that several moves can be done by broadcasting when dest_locations have more than one dimension.
```

File: src/spikeinterface/core/template_tools.py [hunk 7]
```diff
@@ -248,3 +249,165 @@ def get_template_extremum_amplitude(
...
+    templates: np.array
+        A numpy array with dense templates.
+        shape = (num_template, num_sample, num_channel)
+    source_locations: np.array 
+        The channel source location corresponding to templates.
+        shape = (num_channel, 2)
+    dest_locations: np.array
+        The new channel position, if ndim == 3, then the interpolation is broadcated with last dim.
+        shape = (num_channel, 2) or (num_motion, num_channel, 2)
+    interpolation_method: str, default "cubic"
```
Comment (zm711) [inline src/spikeinterface/core/template_tools.py:381]:
```suggestion
    interpolation_method: "cubic" | "linear", default: "cubic"
```

File: src/spikeinterface/core/template_tools.py [hunk 8]
```diff
@@ -248,3 +249,165 @@ def get_template_extremum_amplitude(
...
+    import scipy.interpolate
+
+    source_locations = np.asarray(source_locations)
+    dest_locations = np.asarray(dest_locations)
+    if dest_locations.ndim == 2:
+        new_shape = templates.shape
+    elif dest_locations.ndim == 3:
+        new_shape = (dest_locations.shape[0], ) + templates.shape
+    else:
+        raise ValueError("Bad dest_locations")
```
Comment (zm711) [inline src/spikeinterface/core/template_tools.py:398]:
```suggestion
        raise ValueError(f"Incorrect dimensions for dest_locations: {dest_locations.ndim}. Dimension can be 2 or 3. ")
```

File: src/spikeinterface/core/template_tools.py [hunk 9]
```diff
@@ -248,3 +249,165 @@ def get_template_extremum_amplitude(
...
+    Returns
+    -------
+    new_templates: list of Templates
+
+    """
+    if templates.are_templates_sparse():
+        # TODO make some check!!
+        pass
+    else:
+        warnings.warn("move_templates() with dense templates cannot checks that template can be interpolated at all places")
```
Comment (zm711) [inline src/spikeinterface/core/template_tools.py:290]:
```suggestion
        warnings.warn("move_templates() with dense templates cannot check that templates can be interpolated at all places")
```

I think the better preposition may be `to all places`, but you could interpolate at all places of the vector. So I think `at` if you want to emphasize the interpolation acting on the data vs `to` which emphasizes the result of the interpolation.

**Review by alejoe91 — APPROVED** [review]

**Pull request merged**